tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -4077,63 +4150,42 @@ std::optional<Value> convertGatherOp(PatternRewriter& rewriter, Operation* op,
                                            params_type.getElementType()),
       params_value, params_transpose_perm_val.value());
 
-  if (count_dynamic_dims(tosa_values_shape) > 1) {
-    return (void)rewriter.notifyMatchFailure(
-               op,
-               "multiply dynamic shapes when reshaping values down to "
-               "tosa.gather"),
-           std::nullopt;
+  std::optional<Value> tosa_values_reshape_op = build_reshape(
+      rewriter, op->getLoc(), params_transpose_op, {N, K, C}, tosaOnly);
+
+  if (!tosa_values_reshape_op) {
+    (void)rewriter.notifyMatchFailure(op, "unable to reshape values");
+    return std::nullopt;
   }
 
-  auto tosa_values_reshape_op = CreateOpAndInfer<tosa::ReshapeOp>(
-      rewriter, op->getLoc(),
-      tensorflow::GetTypeFromTFTensorShape(tosa_values_shape,
-                                           params_type.getElementType()),
-      params_transpose_op.getResult(),
-      rewriter.getDenseI64ArrayAttr(
-          tensorflow::ConvertMlirShapeToTF(tosa_values_shape)));
+  std::optional<Value> tosa_indices_reshape_op =
+      build_reshape(rewriter, op->getLoc(), indices_value, {N, W}, tosaOnly);
 
-  if (count_dynamic_dims(tosa_indices_shape) > 1) {
-    return (void)rewriter.notifyMatchFailure(
-               op,
-               "multiply dynamic shapes when reshaping indices down to "
-               "tosa.gather"),
-           std::nullopt;
+  if (!tosa_indices_reshape_op) {
+    (void)rewriter.notifyMatchFailure(op, "unable to reshape indices");
+    return std::nullopt;
```

Comment:
Avoid doing construction check-fails throughout. The pattern rewriters often fail in confusing ways if you construct a larger number of results then `return failure()`. Instead validate that the case is correct, then perform the constructions only if the values look correct.

Reply:
These failure cases disappear after some of the suggested fixes.